tables do not allow setting of metadata=None even if 'null' is allowed for metadata

We made it so that we can allow empty metadata by setting `type = ['object', 'null']` in the metadata schema. If so, then
```
schema.validate_and_encode_row(None)
```
returns `b''`. However, you *can't* actually do
```
table.add_row(metadata=None)
```
because you get the error
```
E           tskit.exceptions.MetadataValidationError:
... 
E           
E           On instance:
E               {}

```

The culprit is right here: https://github.com/tskit-dev/tskit/blob/main/python/tskit/metadata.py#L571
because `metadata=None` gets turned into `{}` here: https://github.com/tskit-dev/tskit/blob/main/python/tskit/tables.py#L549

A test and possible solution to follow.

--- a/python/tests/test_metadata.py
+++ b/python/tests/test_metadata.py
@@ -439,11 +439,19 @@ class TestMetadataModule:
             with pytest.raises(exceptions.MetadataSchemaValidationError):
                 metadata.MetadataSchema(schema)
 
-    def test_null_union_top_level(self):
+    @pytest.mark.parametrize("codec", ["struct", "json"])
+    def test_null_union_top_level(self, codec):
         schema = {
-            "codec": "json",
+            "codec": f"{codec}",
             "type": ["object", "null"],
-            "properties": {"one": {"type": "string"}, "two": {"type": "number"}},
+            "properties": {
+                "one": {
+                    "type": "string",
+                    "binaryFormat": "1024s",
+                    "nullTerminated": True,
+                },
+                "two": {"type": "number", "binaryFormat": "i"},
+            },
         }
         ms = metadata.MetadataSchema(schema)
         row_data = {"one": "tree", "two": 5}
@@ -1025,6 +1033,25 @@ class TestStructCodec:
             "float": 5.5,
             "int": 42,
         }
+
+    def test_defaults_object_or_null(self):
+        schema = {
+            "codec": "struct",
+            "type": ["object", "null"],
+            "properties": {
+                "int": {"type": "number", "binaryFormat": "b", "default": 42},
+                "float": {"type": "number", "binaryFormat": "d"},
+            },
+        }
+        ms = metadata.MetadataSchema(schema)
+        row_data = {"float": 5.5}
+        assert ms.validate_and_encode_row(row_data) == b"\x00\x00\x00\x00\x00\x00\x16@*"
+        assert ms.decode_row(ms.validate_and_encode_row(row_data)) == {
+            "float": 5.5,
+            "int": 42,
+        }
+        assert ms.validate_and_encode_row(None) == b""
+        assert ms.decode_row(b"") is None
 
 
 class TestStructCodecRoundTrip:

--- a/python/tests/test_tables.py
+++ b/python/tests/test_tables.py
@@ -851,6 +851,105 @@ class MetadataTestsMixin:
                 assert table[j].metadata == metadata_column[j]
                 assert table[j + num_rows].metadata == metadata_column[j]
 
+    @pytest.mark.parametrize(
+        "codec",
+        ["struct", "json"],
+    )
+    def test_set_null_metadata(self, codec):
+        table = self.table_class()
+        table.metadata_schema = metadata.MetadataSchema(
+            {
+                "codec": f"{codec}",
+                "title": "Example Metadata",
+                "type": ["object", "null"],
+                "properties": {
+                    "a": {"type": "number", "binaryFormat": "i"},
+                },
+                "required": ["a"],
+                "additionalProperties": False,
+            },
+        )
+        examples = [{"a": 4}, None]
+        for md in examples:
+            table.add_row(
+                **{
+                    **self.input_data_for_add_row(),
+                    "metadata": md,
+                }
+            )
+        assert table.num_rows == len(examples)
+        for md, row in zip(examples, table):
+            assert md == row.metadata
+
+    # only json allows leaving out of optional entries
+    def test_set_empty_metadata_json(self):
+        table = self.table_class()
+        table.metadata_schema = metadata.MetadataSchema(
+            {
+                "codec": "json",
+                "title": "Example Metadata",
+                "type": ["object", "null"],
+                "properties": {
+                    "a": {"type": "number", "binaryFormat": "i"},
+                },
+                "required": [],
+                "additionalProperties": False,
+            },
+        )
+        examples = [{"a": 4}, {}]
+        for md in examples:
+            table.add_row(
+                **{
+                    **self.input_data_for_add_row(),
+                    "metadata": md,
+                }
+            )
+        assert table.num_rows == len(examples)
+        for md, row in zip(examples, table):
+            assert md == row.metadata
+
+    @pytest.mark.parametrize(
+        "codec",
+        [
+            "struct",
+            pytest.param(
+                "json",
+                marks=pytest.mark.xfail(reason="JSON defaults not yet implemented"),
+            ),
+        ],
+    )
+    def test_set_with_optional_properties(self, codec):
+        table = self.table_class()
+        table.metadata_schema = metadata.MetadataSchema(
+            {
+                "codec": f"{codec}",
+                "title": "Example Metadata",
+                "type": ["object", "null"],
+                "properties": {
+                    "a": {"type": "number", "binaryFormat": "i", "default": 0},
+                },
+                "additionalProperties": False,
+            },
+        )
+        metadata_list = [{"a": 4}, None, {"a": 5}, {}]
+        for md in metadata_list:
+            table.add_row(
+                **{
+                    **self.input_data_for_add_row(),
+                    "metadata": md,
+                }
+            )
+        assert table.num_rows == len(metadata_list)
+        for md, row in zip(metadata_list, table):
+            # If None is allowed by the schema it gets used even in the presence of
+            # default values.
+            if isinstance(md, dict):
+                defaults = {"a": 0}
+                defaults.update(md)
+                assert defaults == row.metadata
+            else:
+                assert md == row.metadata
+
 
 class TestIndividualTable(CommonTestsMixin, MetadataTestsMixin):
     columns = [UInt32Column("flags")]

--- a/python/tskit/metadata.py
+++ b/python/tskit/metadata.py
@@ -446,13 +446,23 @@ class StructCodec(AbstractMetadataCodec):
             key: StructCodec.make_encode(prop)
             for key, prop in sub_schema["properties"].items()
         }
-        return (
-            lambda obj: b""
-            if obj is None
-            else b"".join(
-                sub_encoder(obj[key]) for key, sub_encoder in sub_encoders.items()
-            )
-        )
+        defaults = {
+            key: prop["default"]
+            for key, prop in sub_schema["properties"].items()
+            if "default" in prop
+        }
+
+        def object_encode(obj):
+            values = []
+            if obj is not None:
+                for key, sub_encoder in sub_encoders.items():
+                    try:
+                        values.append(sub_encoder(obj[key]))
+                    except KeyError:
+                        values.append(sub_encoder(defaults[key]))
+            return b"".join(values)
+
+        return object_encode
 
     @classmethod
     def make_string_encode(cls, sub_schema):
@@ -568,7 +578,12 @@ class MetadataSchema:
             self._validate_row = TSKITMetadataSchemaValidator(schema).validate
             self.encode_row = codec_instance.encode
             self.decode_row = codec_instance.decode
-            self.empty_value = {}
+            # If None is allowed by the schema it gets used even in the presence of
+            # default and required values.
+            if "type" in schema and "null" in schema["type"]:
+                self.empty_value = None
+            else:
+                self.empty_value = {}
 
     def __repr__(self) -> str:
         return self._string

